java/rocksjni/rocksjni.cc

```diff
@@ -746,6 +747,282 @@ void Java_org_rocksdb_RocksDB_putDirect(
+    const ROCKSDB_NAMESPACE::WriteOptions default_write_options =
+        ROCKSDB_NAMESPACE::WriteOptions();
+
+    auto columns_family =
+        (jcf_handle == 0)
+            ? db->DefaultColumnFamily()
+            : reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyHandle*>(
+                  jcf_handle);
+
+    db->PutEntity(default_write_options, columns_family, key, columns);
+  };
+
+  ROCKSDB_NAMESPACE::WideColumnJni::convertWideColumnsDirect(
+      putEntity, env, jKey, key_offset, key_len, jnames, jnames_offset,
+      jnames_len, jvalues, jvalues_offset, jvalues_len);
+}
+
+/*
+ * Class:     org_rocksdb_RocksDB
+ * Method:    getEntityDirect
+ * Signature:
+ * (JLjava/nio/ByteBuffer;II[Ljava/nio/ByteBuffer;[I[I[I[Ljava/nio/ByteBuffer;[I[I[I)Lorg/rocksdb/Status;
+ */
+jobject Java_org_rocksdb_RocksDB_getEntityDirect(
+    JNIEnv* env, jobject, jlong jdb_handle, jobject jKey, jint key_offset,
+    jint key_len, jobjectArray jnames, jintArray jNamesOffset,
+    jintArray jNamesLen, jintArray jnamesRequiredSize, jobjectArray jvalues,
+    jintArray jValuesOffset, jintArray jValuesLen,
+    jintArray jvaluesRequiredSize, jlong jcf_handle) {
+  auto* db = reinterpret_cast<ROCKSDB_NAMESPACE::DB*>(jdb_handle);
+  auto columns_family =
+      (jcf_handle == 0)
+          ? db->DefaultColumnFamily()
+          : reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyHandle*>(
+                jcf_handle);
+
+  auto _key = env->GetDirectBufferAddress(jKey);
+  if (_key == nullptr) {
+    ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+        env,
+        "Invalid key argument (argument is not a valid direct ByteBuffer)");
+    return nullptr;
+  }
+  auto key = reinterpret_cast<char*>(_key) + key_offset;
+
+  ROCKSDB_NAMESPACE::Slice key_slice(key, static_cast<size_t>(key_len));
+
+  const ROCKSDB_NAMESPACE::ReadOptions default_read_options =
+      ROCKSDB_NAMESPACE::ReadOptions();
+  ROCKSDB_NAMESPACE::PinnableWideColumns columns;
+
+  const auto status =
+      db->GetEntity(default_read_options, columns_family, key_slice, &columns);
+
+  if (status.ok() && columns.columns().size() > 0) {
+    auto max_items = std::min(static_cast<jsize>(columns.columns().size()),
+                              env->GetArrayLength(jnames));
+
+    auto namesOffset = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jNamesOffset, 0, max_items, namesOffset.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto namesLen = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jNamesLen, 0, max_items, namesLen.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto namesRequiredSize = std::make_unique<jint[]>(max_items);
+
+    auto valuesOffset = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jValuesOffset, 0, max_items, valuesOffset.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
```

Comment:
Do we need to free the previous JNI array fetches here?

Reply:
No we don't need to. See previous comment